dttools/src/copy_tree.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (C) 2013- The University of Notre Dame
+ * This software is distributed under the GNU General Public License.
+ * See the file COPYING for details.
+ */
+
+#include <assert.h>
+#include <dirent.h>
+#include <errno.h>
+#include <libgen.h>
+#include <limits.h>
+#include <string.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "copy_tree.h"
+#include "copy_stream.h"
+#include "create_dir.h"
+#include "debug.h"
+#include "path.h"
+#include "xxmalloc.h"
+
+mode_t default_dirmode = S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH;
+
+int copy_dir_real(const char *source, const char *target);
+
+int copy_symlink(const char *source, const char *target) {
+	char linkname[PATH_MAX];
+	ssize_t r;
+
+	assert(source);
+	assert(target);
+
+	if(!access(target, F_OK)) {
+		LDEBUG("%s already exists!\n", target);
+		return -1;
+	}
+
+	r = readlink(source, linkname, sizeof(linkname));
+	if(r == -1) {
+		LDEBUG("readlink(`%s`) failed: %s!\n", source, strerror(errno));
+		return -1;
+	}
+	linkname[r] = '\0';
+
+	if(symlink(linkname, target)) {
+		LDEBUG("symlink(`%s`, `%s`) failed: %s\n", linkname, target, strerror(errno));
```

Comment:
Same reason with above. 
`LDEBUG` is a debug-level info, and used in all the codes I added.